src/Revit/RevitHyparTools/RevitExtensions.cs

```diff
@@ -1,11 +1,90 @@
+using System.Collections.Generic;
+using System.Linq;
 using Autodesk.Revit.DB;
 using Elements.Geometry;
 
-namespace Hypar.Revit 
+namespace Hypar.Revit
 {
-    public static class RevitExtensions {
-        public static Vector3 ToVector3(this XYZ xyz) {
-            return new Vector3(xyz.X, xyz.Y, xyz.Z);
-        }       
+    internal static class RevitExtensions
+    {
+        /// <summary>
+        /// Convert the Revit XYZ to an Elements Vector3, with an option to scale the 
+        /// Vector3 from feet to meters.  This scaling should be done with any XYZ that represents
+        /// a geometric property, but should maybe be avoided for other vectors.
+        /// </summary>
+        internal static Vector3 ToVector3(this XYZ xyz, bool scaleToMeters = false)
+        {
+            var scale = 1.0;
+            if (scaleToMeters)
+            {
+                scale = Elements.Units.FeetToMeters(1);
+            }
+            return new Vector3(xyz.X, xyz.Y, xyz.Z) * scale;
+        }
+
+        /// <summary>
+        /// Retrieve all of the Element.Geometry.Profiles that describe the Revit PlanarFace
+        /// Soemtimes a single PlanarFace in Revit and actually multiple "faces" should be translated into multiple Element Profiles
+        /// </summary>
+        internal static Elements.Geometry.Profile[] GetProfiles(this PlanarFace face)
+        {
+            var polygons = face.GetEdgesAsCurveLoops().Select(cL => cL.ToPolygon());
+
+            var polygonLoopDict = MatchOuterLoopPolygonsWithInnerHoles(polygons);
+
+            var profiles = polygonLoopDict.Select(kvp => new Elements.Geometry.Profile(kvp.Key, kvp.Value, Guid.NewGuid(), "Revit Profile"));
+            var scaledProfiles = profiles.Select(p => p.Scale(Elements.Units.FeetToMeters(1)));
+            return scaledProfiles.ToArray();
+        }
+
+        /// <summary>
+        /// Analyze all of the PlanarFaces of a Revit solid and returns those that 
+        /// face "up" within a certain threshold.
+        /// </summary>
+        /// <param name="solid">The revit Solid</param>
+        /// <param name="verticalThreshold">The angle (in degrees) that represents the threshold for a face to be considered facing up.  A completely horizontal face will have an angle of 0.  A face that does not face up or down at all will have an angle of 90.</param>
+        private static PlanarFace[] GetMostLikelyTopFaces(this Solid solid, double verticalThreshold = 30)
+        {
+            var faces = new List<PlanarFace>();
+            foreach (PlanarFace face in solid.Faces)
+            {
+                if (face.FaceNormal.DotProduct(XYZ.BasisZ) > Math.Cos(Elements.Units.DegreesToRadians(verticalThreshold)) && face.FaceNormal.DotProduct(XYZ.BasisZ) <= 1)
+                {
+                    faces.Add(face);
+                }
+            }
+            return faces.ToArray();
+        }
+
+        private static Polygon ToPolygon(this CurveLoop curveLoop)
```

Comment:
Can we add a scaling parameter here as well? Then push it down to the call to `ToVector3()`? This will save us from the additional scale operation you do above in `GetProfiles(...)`

Reply:
sounds good.